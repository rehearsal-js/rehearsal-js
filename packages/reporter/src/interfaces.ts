/*
! THE REPORTER OUTPUT IS BOTH A TERMINAL PRINT AND A JSON FILE. SINCE WE DONT WANT _ KEYS IN THE JSON FILE, WE USE CAMEL
*/

import type { SourceLocation } from "jscodeshift";

export interface ReporterOptions {
  cwd?: string;
  filename?: string;
  projectName?: string;
  tscVersion?: string;
}

export interface Report extends ReporterSummary {
  projectName: string;
  fileCount: number;
  tscVersion: string;
  tscLog: TSCLog[];
}

export interface TSCLog extends ReporterSummary {
  filePath: string;
  errors: TSCLogError[];
}

export interface ReporterSummary {
  cumulativeErrors: number;
  uniqueErrors: number;
  uniqueErrorList: string[];
  autofixedCumulativeErrors: number;
  autofixedErrorList: string[];
}

export interface TSCLogError {
  errorCode: string;
  errorCategory: string;
  errorMessage: string;
  sourceLocation: SourceLocation;
  source: string;
  helpMessage: string;
}
