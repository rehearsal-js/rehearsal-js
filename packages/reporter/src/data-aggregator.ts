import ts from 'typescript';
import { FixResult, ProcessedFile, FileRole } from './types';

export class DataAggregator {
  private static instance: DataAggregator;
  static getInstance(diagnostic: ts.DiagnosticWithLocation): DataAggregator {
    if(!DataAggregator.instance || DataAggregator.instance.diagnostic !== diagnostic ) {
      DataAggregator.instance = new DataAggregator(diagnostic);
    }
    return DataAggregator.instance;
  }
  
  private result: FixResult;
  private diagnostic: ts.DiagnosticWithLocation;
  
  constructor(diagnostic: ts.DiagnosticWithLocation) {
    this.diagnostic = diagnostic;
    this.result = this.getInitialResult(this.diagnostic);
  }

  getInitialResult(diagnostic: ts.DiagnosticWithLocation): FixResult {
    const { file, start } = diagnostic;
    const location = this.getLocation(file, start);
    return {
      analysisTarget: file.fileName,
      files: {
        [file.fileName]: {
          ...this.buildFile(file.fileName),
          location,
          roles: ['analysisTarget', 'unmodified'],
        },
      },
      fixed: false,
      hintAdded: false,
    };
  }

  addCommentDataToResult(
    fileName: string,
    hint = '',
    roles: FileRole[] = [],
    location?: { line: number | undefined; character: number | undefined }
  ): void {
    if(this.result) {
      const file = fileName in this.result.files ? this.result.files[fileName] : this.buildFile(fileName);
      this.result.files[fileName] = {
        ...file,
        hintAdded: true,
        location: location || file.location,
        hint,
        roles: this.mergeRoles(file.roles, roles),
      };
      this.result = {
        ...this.result,
        hintAdded: true
      }
    };
  }

  addCodemodDataToResult(
    fileName: string,
    updatedText = '',
    code = '',
    roles: FileRole[] = [],
    location?: { line: number | undefined; character: number | undefined },
  ): void {
    if(this.result) {
      const file = fileName in this.result.files ? this.result.files[fileName] : this.buildFile(fileName);
      this.result.files[fileName] = {
        ...file,
        fixed: true,
        updatedText,
        location: location || file.location,
        code,
        roles: this.mergeRoles(file.roles, roles),
      };

      this.result = {
        ...this.result,
        fixed: true,
      }
    }
  }

  getResult(): FixResult{
    return this.result;
  }

  private buildFile(fileName: string): ProcessedFile {
    return {
      fileName,
      updatedText: '',
      location: {
        line: undefined,
        character: undefined,
      },
      fixed: false,
      code: '',
      hintAdded: false,
      hint: '',
      roles: [],
    };
  }

  private mergeRoles(existingRoles: FileRole[], newRoles: FileRole[]): FileRole[] {
    const roles = Array.from(new Set([...existingRoles, ...newRoles, 'modified' as const])).filter(
      (role) => role !== 'unmodified'
    );
    return roles;
  }

  getLocation(sourceFile: ts.SourceFile, position: number) {
    const { line, character } = ts.getLineAndCharacterOfPosition(sourceFile, position);
    return {
      line,
      character
    };
  }
}