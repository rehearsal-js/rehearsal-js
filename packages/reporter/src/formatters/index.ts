export * from './json-formatter.js';
export * from './md-formatter.js';
export * from './sonarqube-formatter.js';
export * from './sarif-formatter.js';
