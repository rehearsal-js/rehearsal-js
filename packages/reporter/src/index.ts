export { Reporter } from './reporter';
export { jsonFormatter } from './formatters/json-formatter';
export { mdFormatter } from './formatters/md-formatter';
export { sarifFormatter } from './formatters/sarif-formatter';
export { sonarqubeFormatter } from './formatters/sonarqube-formatter';
export { normalizeFilePath } from './normalize-paths';

export type { Location, Report, ReportItem, ReportSummary, ReportFormatter } from './types';
