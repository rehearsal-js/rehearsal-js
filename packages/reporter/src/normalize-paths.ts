import { isAbsolute, relative } from 'node:path';

export function normalizeFilePath(basePath: string, filepath: string): string {
  if (isAbsolute(filepath)) {
    return relative(basePath, filepath);
  }

  return filepath;
}
