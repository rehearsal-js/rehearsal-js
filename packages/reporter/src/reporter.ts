import { existsSync, readFileSync, writeFileSync } from 'fs';
import { DiagnosticCategory, flattenDiagnosticMessageText, SyntaxKind } from 'typescript';
import {
  type ProcessedFile,
  type Report,
  type ReportFormatter,
  type ReportItem,
  type Location,
} from './types';
import type { DiagnosticWithLocation, Node } from 'typescript';
import type { Logger } from 'winston';

type ReporterMeta = {
  projectName: string;
  basePath: string;
  commandName: string;
  tsVersion: string;
};

/**
 * Representation of diagnostic and migration report.
 */
export class Reporter {
  readonly basePath: string;

  public report: Report;
  private logger?: Logger;

  constructor(meta: ReporterMeta, logger?: Logger) {
    const { projectName, basePath, commandName, tsVersion } = meta;

    this.basePath = basePath;
    this.logger = logger?.child({ service: 'rehearsal-reporter' });

    this.report = {
      summary: {
        projectName: projectName,
        tsVersion: tsVersion,
        timestamp: new Date().toLocaleString('en-US', {
          year: 'numeric',
          month: 'numeric',
          day: 'numeric',
          hourCycle: 'h24',
          hour: '2-digit',
          minute: '2-digit',
          second: '2-digit',
        }),
        basePath: basePath,
        commandName: commandName,
      },
      items: [],
    };
  }

  getFileNames(): string[] {
    return [...new Set(this.report.items.map((item) => item.analysisTarget))];
  }

  getItemsByAnalysisTarget(fileName: string): ReportItem[] {
    return this.report.items.filter((item) => item.analysisTarget === fileName);
  }

  /**
   * Appends an element to the summary
   */
  addSummary(key: string, value: unknown): void {
    this.report.summary[key] = value;
  }

  /**
   * Appends am information about provided diagnostic and related node to the report
   */
  addItem(
    diagnostic: DiagnosticWithLocation,
    files: { [fileName: string]: ProcessedFile },
    fixed: boolean,
    node?: Node,
    triggeringLocation?: Location,
    hint = '',
    helpUrl = ''
  ): void {
    this.report.items.push({
      analysisTarget: diagnostic.file.fileName,
      files,
      errorCode: diagnostic.code,
      category: DiagnosticCategory[diagnostic.category],
      message: flattenDiagnosticMessageText(diagnostic.messageText, '. '),
      hint: hint,
      fixed,
      nodeKind: node ? SyntaxKind[node.kind] : undefined,
      nodeText: node?.getText(),
      helpUrl,
      //Original node that triggers the error; may not be the node where the codefix happens
      //Location in type ProcessedFile is where the codefix happens or the message is added
      //The two locations in most cases will be the same, but can be different
      nodeLocation: triggeringLocation || undefined,
    });
  }

  /**
   * Prints the current report using provided formatter (ex. json, pull-request etc.)
   */
  print(file: string, formatter: ReportFormatter): string {
    const report = formatter(this.report);

    if (file) {
      writeFileSync(file, report);
    }

    return report;
  }

  /**
   * Saves the current report information to the file in simple JSON format
   * to be able to load it later with 'load' function
   */
  save(file: string): void {
    this.print(file, (report: Report): string => JSON.stringify(report, null, 2));
    this.logger?.info(`Report saved to: ${file}.`);
  }

  /**
   * Loads the report exported by function 'save' from the file
   */
  load(file: string): Reporter {
    if (!existsSync(file)) {
      this.logger?.error(`Report file not found: ${file}.`);
    }

    this.logger?.info(`Report file found: ${file}.`);
    const content = readFileSync(file, 'utf-8');
    const report: Report = JSON.parse(content);

    if (!Reporter.isReport(report)) {
      this.logger?.error(`Report not loaded: wrong file format`);
      return this;
    }

    this.report = report as Report;
    this.logger?.info(`Report loaded from file.`);

    return this;
  }

  private static isReport(report: Report): report is Report {
    return report && report.summary !== undefined && report.items !== undefined;
  }
}
