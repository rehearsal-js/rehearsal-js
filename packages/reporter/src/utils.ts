import { readFileSync } from 'fs';
import { parse } from 'json5';

export function readText(file: string): string | undefined {
  return readFile(file, 'utf8');
}

export function readFile(file: string, encoding: 'utf8'): string | undefined;
export function readFile(file: string, encoding?: undefined): Buffer | undefined;
export function readFile(file: string, encoding?: 'utf8'): string | Buffer | undefined {
  try {
    return readFileSync(file, encoding);
  } catch (e) {
    if ((e as NodeJS.ErrnoException).code === 'ENOENT') {
      return;
    }
    throw e;
  }
}

export function readJSONString<TJson = unknown>(filepath: string, separator = '\n'): TJson[] {
  const text = readText(filepath) as string;
  const strJSONArray = text.split(separator).filter((a) => a);

  return strJSONArray.map((a) => parse(a));
}

export async function sleep(ms: number): Promise<void> {
  return new Promise((resolve) => setTimeout(resolve, ms));
}
