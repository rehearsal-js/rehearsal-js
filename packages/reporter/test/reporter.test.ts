import { assert } from "chai";
import { suite } from "mocha";
import { dirSync, setGracefulCleanup } from "tmp";
import { join } from "path";
import { existsSync, readJSONSync } from "fs-extra";

import { Reporter } from "../src";

import type {
  TSCLog,
  Report
} from "../src/interfaces";

const TMP_DIR = dirSync().name;
const RESULTS_FILEPATH = join(TMP_DIR, ".rehearsal.json");
const REPORTER = new Reporter({cwd: TMP_DIR});
const PROJECT_NAME = "my super amazing project";
const TSC_VERSION = "0.0.1-dev.20211116";
const ENTRY_FILEPATH = "src/foo.ts";
const ENTRY_ERROR_CODE = "TS1354";

const LOG_ENTRY: TSCLog = {
  filePath: ENTRY_FILEPATH,
  cumulativeErrors: 0,
  uniqueErrors: 0,
  uniqueErrorList: [],
  autofixedCumulativeErrors: 0,
  autofixedErrorList: [],
  errors: [{
    errorCode: ENTRY_ERROR_CODE,
    errorCategory: "error",
    errorMessage: "error message is foo la la",
    helpMessage: "this will help you",
    stringLocation: {
      start: 0,
      end: 0
    }
  }],
};

REPORTER.projectName = PROJECT_NAME;
REPORTER.tscVersion = TSC_VERSION;
REPORTER.terminalLogger.error(`Rehearsal: Your on your own.`);
REPORTER.fileLogger.log("info", "tsc-log-entry", LOG_ENTRY);

suite("reporter", () => {
  it(`runs reporter.end()`, async () => {
    await REPORTER.end();
    assert.ok(
      existsSync(RESULTS_FILEPATH),
      `result file ${RESULTS_FILEPATH} should exists`
    );

    const report: Report = readJSONSync(RESULTS_FILEPATH);

    assert.equal(report.projectName, PROJECT_NAME);
    assert.equal(report.tscVersion, TSC_VERSION);
    assert.equal(report.tscLog[0].filePath, ENTRY_FILEPATH);
    assert.deepEqual(report.uniqueErrorList, [ENTRY_ERROR_CODE]);
    assert.equal(report.uniqueErrors, 1);
    assert.equal(report.cumulativeErrors, 1);
    assert.equal(report.fileCount, 1);
  });
}).afterEach(() => {
  setGracefulCleanup();
});
