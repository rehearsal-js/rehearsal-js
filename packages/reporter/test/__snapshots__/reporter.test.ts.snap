// Vitest Snapshot v1

exports[`Test reporter > addItem 1`] = `
"{
  \\"summary\\": {
    \\"projectName\\": \\"@rehearsal/test\\",
    \\"tsVersion\\": \\"4.7.4\\",
    \\"timestamp\\": \\"9/16/2022, 13:24:57\\",
    \\"basePath\\": \\"/base/path\\",
    \\"commandName\\": \\"@rehearsal/reporter\\"
  },
  \\"items\\": [
    {
      \\"analysisTarget\\": \\"/base/path/file1.ts\\",
      \\"files\\": {
        \\"/base/path/file1.ts\\": {
          \\"fileName\\": \\"/base/path/file1.ts\\",
          \\"location\\": {
            \\"startLine\\": 1,
            \\"startColumn\\": 1,
            \\"endLine\\": 1,
            \\"endColumn\\": 1
          },
          \\"fixed\\": true,
          \\"code\\": \\"import react from 'react';\\",
          \\"codeFixAction\\": \\"delete\\",
          \\"hintAdded\\": false,
          \\"roles\\": [
            \\"analysisTarget\\",
            \\"modified\\"
          ]
        }
      },
      \\"errorCode\\": 6133,
      \\"category\\": \\"Error\\",
      \\"message\\": \\"'react' is declared but its value is never read.\\",
      \\"hint\\": \\"The declaration 'react' is never read or used. Remove the declaration or use it.\\",
      \\"fixed\\": true,
      \\"nodeKind\\": \\"ImportDeclaration\\",
      \\"nodeText\\": \\"import react from 'react';\\",
      \\"nodeLocation\\": {
        \\"startLine\\": 1,
        \\"startColumn\\": 1,
        \\"endLine\\": 1,
        \\"endColumn\\": 12
      }
    },
    {
      \\"analysisTarget\\": \\"/base/path/file1.ts\\",
      \\"files\\": {
        \\"/base/path/file1.ts\\": {
          \\"fileName\\": \\"/base/path/file1.ts\\",
          \\"location\\": {
            \\"startLine\\": 1,
            \\"startColumn\\": 10,
            \\"endLine\\": 1,
            \\"endColumn\\": 20
          },
          \\"fixed\\": false,
          \\"hint\\": \\"The variable 'a' has type 'number', but 'string' is assigned. Please convert 'string' to 'number' or change variable's type.\\",
          \\"hintAdded\\": true,
          \\"roles\\": [
            \\"analysisTarget\\",
            \\"unmodified\\"
          ]
        }
      },
      \\"errorCode\\": 2322,
      \\"category\\": \\"Error\\",
      \\"message\\": \\"Type 'string' is not assignable to type 'number'.\\",
      \\"hint\\": \\"The variable 'a' has type 'number', but 'string' is assigned. Please convert 'string' to 'number' or change variable's type.\\",
      \\"fixed\\": false,
      \\"nodeKind\\": \\"Identifier\\",
      \\"nodeText\\": \\"a\\",
      \\"nodeLocation\\": {
        \\"startLine\\": 1,
        \\"startColumn\\": 10,
        \\"endLine\\": 1,
        \\"endColumn\\": 20
      }
    },
    {
      \\"analysisTarget\\": \\"/base/path/file1.ts\\",
      \\"files\\": {
        \\"/base/path/file1.ts\\": {
          \\"fileName\\": \\"/base/path/file1.ts\\",
          \\"location\\": {
            \\"startLine\\": 1,
            \\"startColumn\\": 10,
            \\"endLine\\": 1,
            \\"endColumn\\": 26
          },
          \\"fixed\\": false,
          \\"hint\\": \\"Argument of type '{0}' is not assignable to parameter of type 'string'. Consider verifying both types, using type assertion: '({} as string)', or using type guard: 'if ({} instanceof string) { ... }'.\\",
          \\"hintAdded\\": true,
          \\"roles\\": [
            \\"analysisTarget\\",
            \\"unmodified\\"
          ]
        }
      },
      \\"errorCode\\": 2345,
      \\"category\\": \\"Error\\",
      \\"message\\": \\"Argument of type '{}' is not assignable to parameter of type 'string'.\\",
      \\"hint\\": \\"Argument of type '{0}' is not assignable to parameter of type 'string'. Consider verifying both types, using type assertion: '({} as string)', or using type guard: 'if ({} instanceof string) { ... }'.\\",
      \\"fixed\\": false,
      \\"nodeKind\\": \\"ObjectLiteralExpression\\",
      \\"nodeText\\": \\"{}\\",
      \\"nodeLocation\\": {
        \\"startLine\\": 1,
        \\"startColumn\\": 10,
        \\"endLine\\": 1,
        \\"endColumn\\": 26
      }
    },
    {
      \\"analysisTarget\\": \\"testFile1.ts\\",
      \\"files\\": {
        \\"/base/path/testFile.ts\\": {
          \\"fileName\\": \\"/base/path/testFile.ts\\",
          \\"location\\": {
            \\"startLine\\": 3,
            \\"startColumn\\": 7,
            \\"endLine\\": 3,
            \\"endColumn\\": 12
          },
          \\"fixed\\": false,
          \\"hint\\": \\"This is the hint.\\",
          \\"hintAdded\\": true,
          \\"roles\\": [
            \\"analysisTarget\\",
            \\"unmodified\\"
          ]
        }
      },
      \\"errorCode\\": 1000,
      \\"category\\": \\"Error\\",
      \\"message\\": \\"unused variable\\",
      \\"hint\\": \\"\\",
      \\"fixed\\": false,
      \\"helpUrl\\": \\"\\",
      \\"nodeLocation\\": {
        \\"startLine\\": 0,
        \\"startColumn\\": 5,
        \\"endLine\\": 0,
        \\"endColumn\\": 5
      }
    }
  ]
}"
`;

exports[`Test reporter > getItemsByFile 1`] = `
[
  {
    "analysisTarget": "/base/path/file1.ts",
    "category": "Error",
    "errorCode": 6133,
    "files": {
      "/base/path/file1.ts": {
        "code": "import react from 'react';",
        "codeFixAction": "delete",
        "fileName": "/base/path/file1.ts",
        "fixed": true,
        "hintAdded": false,
        "location": {
          "endColumn": 1,
          "endLine": 1,
          "startColumn": 1,
          "startLine": 1,
        },
        "roles": [
          "analysisTarget",
          "modified",
        ],
      },
    },
    "fixed": true,
    "hint": "The declaration 'react' is never read or used. Remove the declaration or use it.",
    "message": "'react' is declared but its value is never read.",
    "nodeKind": "ImportDeclaration",
    "nodeLocation": {
      "endColumn": 12,
      "endLine": 1,
      "startColumn": 1,
      "startLine": 1,
    },
    "nodeText": "import react from 'react';",
  },
  {
    "analysisTarget": "/base/path/file1.ts",
    "category": "Error",
    "errorCode": 2322,
    "files": {
      "/base/path/file1.ts": {
        "fileName": "/base/path/file1.ts",
        "fixed": false,
        "hint": "The variable 'a' has type 'number', but 'string' is assigned. Please convert 'string' to 'number' or change variable's type.",
        "hintAdded": true,
        "location": {
          "endColumn": 20,
          "endLine": 1,
          "startColumn": 10,
          "startLine": 1,
        },
        "roles": [
          "analysisTarget",
          "unmodified",
        ],
      },
    },
    "fixed": false,
    "hint": "The variable 'a' has type 'number', but 'string' is assigned. Please convert 'string' to 'number' or change variable's type.",
    "message": "Type 'string' is not assignable to type 'number'.",
    "nodeKind": "Identifier",
    "nodeLocation": {
      "endColumn": 20,
      "endLine": 1,
      "startColumn": 10,
      "startLine": 1,
    },
    "nodeText": "a",
  },
  {
    "analysisTarget": "/base/path/file1.ts",
    "category": "Error",
    "errorCode": 2345,
    "files": {
      "/base/path/file1.ts": {
        "fileName": "/base/path/file1.ts",
        "fixed": false,
        "hint": "Argument of type '{0}' is not assignable to parameter of type 'string'. Consider verifying both types, using type assertion: '({} as string)', or using type guard: 'if ({} instanceof string) { ... }'.",
        "hintAdded": true,
        "location": {
          "endColumn": 26,
          "endLine": 1,
          "startColumn": 10,
          "startLine": 1,
        },
        "roles": [
          "analysisTarget",
          "unmodified",
        ],
      },
    },
    "fixed": false,
    "hint": "Argument of type '{0}' is not assignable to parameter of type 'string'. Consider verifying both types, using type assertion: '({} as string)', or using type guard: 'if ({} instanceof string) { ... }'.",
    "message": "Argument of type '{}' is not assignable to parameter of type 'string'.",
    "nodeKind": "ObjectLiteralExpression",
    "nodeLocation": {
      "endColumn": 26,
      "endLine": 1,
      "startColumn": 10,
      "startLine": 1,
    },
    "nodeText": "{}",
  },
]
`;

exports[`Test reporter > load 1`] = `
{
  "items": [
    {
      "analysisTarget": "/base/path/file1.ts",
      "category": "Error",
      "errorCode": 6133,
      "files": {
        "/base/path/file1.ts": {
          "code": "import react from 'react';",
          "codeFixAction": "delete",
          "fileName": "/base/path/file1.ts",
          "fixed": true,
          "hintAdded": false,
          "location": {
            "endColumn": 1,
            "endLine": 1,
            "startColumn": 1,
            "startLine": 1,
          },
          "roles": [
            "analysisTarget",
            "modified",
          ],
        },
      },
      "fixed": true,
      "hint": "The declaration 'react' is never read or used. Remove the declaration or use it.",
      "message": "'react' is declared but its value is never read.",
      "nodeKind": "ImportDeclaration",
      "nodeLocation": {
        "endColumn": 12,
        "endLine": 1,
        "startColumn": 1,
        "startLine": 1,
      },
      "nodeText": "import react from 'react';",
    },
    {
      "analysisTarget": "/base/path/file1.ts",
      "category": "Error",
      "errorCode": 2322,
      "files": {
        "/base/path/file1.ts": {
          "fileName": "/base/path/file1.ts",
          "fixed": false,
          "hint": "The variable 'a' has type 'number', but 'string' is assigned. Please convert 'string' to 'number' or change variable's type.",
          "hintAdded": true,
          "location": {
            "endColumn": 20,
            "endLine": 1,
            "startColumn": 10,
            "startLine": 1,
          },
          "roles": [
            "analysisTarget",
            "unmodified",
          ],
        },
      },
      "fixed": false,
      "hint": "The variable 'a' has type 'number', but 'string' is assigned. Please convert 'string' to 'number' or change variable's type.",
      "message": "Type 'string' is not assignable to type 'number'.",
      "nodeKind": "Identifier",
      "nodeLocation": {
        "endColumn": 20,
        "endLine": 1,
        "startColumn": 10,
        "startLine": 1,
      },
      "nodeText": "a",
    },
    {
      "analysisTarget": "/base/path/file1.ts",
      "category": "Error",
      "errorCode": 2345,
      "files": {
        "/base/path/file1.ts": {
          "fileName": "/base/path/file1.ts",
          "fixed": false,
          "hint": "Argument of type '{0}' is not assignable to parameter of type 'string'. Consider verifying both types, using type assertion: '({} as string)', or using type guard: 'if ({} instanceof string) { ... }'.",
          "hintAdded": true,
          "location": {
            "endColumn": 26,
            "endLine": 1,
            "startColumn": 10,
            "startLine": 1,
          },
          "roles": [
            "analysisTarget",
            "unmodified",
          ],
        },
      },
      "fixed": false,
      "hint": "Argument of type '{0}' is not assignable to parameter of type 'string'. Consider verifying both types, using type assertion: '({} as string)', or using type guard: 'if ({} instanceof string) { ... }'.",
      "message": "Argument of type '{}' is not assignable to parameter of type 'string'.",
      "nodeKind": "ObjectLiteralExpression",
      "nodeLocation": {
        "endColumn": 26,
        "endLine": 1,
        "startColumn": 10,
        "startLine": 1,
      },
      "nodeText": "{}",
    },
  ],
  "summary": {
    "basePath": "/base/path",
    "commandName": "@rehearsal/reporter",
    "projectName": "@rehearsal/test",
    "timestamp": "9/16/2022, 13:24:57",
    "tsVersion": "4.7.4",
  },
}
`;

exports[`Test reporter > print, json 1`] = `
"{
  \\"summary\\": {
    \\"projectName\\": \\"@rehearsal/test\\",
    \\"tsVersion\\": \\"4.7.4\\",
    \\"timestamp\\": \\"9/16/2022, 13:24:57\\",
    \\"basePath\\": \\"/base/path\\",
    \\"commandName\\": \\"@rehearsal/reporter\\"
  },
  \\"items\\": [
    {
      \\"analysisTarget\\": \\"/base/path/file1.ts\\",
      \\"files\\": {
        \\"/base/path/file1.ts\\": {
          \\"fileName\\": \\"/base/path/file1.ts\\",
          \\"location\\": {
            \\"startLine\\": 1,
            \\"startColumn\\": 1,
            \\"endLine\\": 1,
            \\"endColumn\\": 1
          },
          \\"fixed\\": true,
          \\"code\\": \\"import react from 'react';\\",
          \\"codeFixAction\\": \\"delete\\",
          \\"hintAdded\\": false,
          \\"roles\\": [
            \\"analysisTarget\\",
            \\"modified\\"
          ]
        }
      },
      \\"errorCode\\": 6133,
      \\"category\\": \\"Error\\",
      \\"message\\": \\"'react' is declared but its value is never read.\\",
      \\"hint\\": \\"The declaration 'react' is never read or used. Remove the declaration or use it.\\",
      \\"fixed\\": true,
      \\"nodeKind\\": \\"ImportDeclaration\\",
      \\"nodeText\\": \\"import react from 'react';\\",
      \\"nodeLocation\\": {
        \\"startLine\\": 1,
        \\"startColumn\\": 1,
        \\"endLine\\": 1,
        \\"endColumn\\": 12
      }
    },
    {
      \\"analysisTarget\\": \\"/base/path/file1.ts\\",
      \\"files\\": {
        \\"/base/path/file1.ts\\": {
          \\"fileName\\": \\"/base/path/file1.ts\\",
          \\"location\\": {
            \\"startLine\\": 1,
            \\"startColumn\\": 10,
            \\"endLine\\": 1,
            \\"endColumn\\": 20
          },
          \\"fixed\\": false,
          \\"hint\\": \\"The variable 'a' has type 'number', but 'string' is assigned. Please convert 'string' to 'number' or change variable's type.\\",
          \\"hintAdded\\": true,
          \\"roles\\": [
            \\"analysisTarget\\",
            \\"unmodified\\"
          ]
        }
      },
      \\"errorCode\\": 2322,
      \\"category\\": \\"Error\\",
      \\"message\\": \\"Type 'string' is not assignable to type 'number'.\\",
      \\"hint\\": \\"The variable 'a' has type 'number', but 'string' is assigned. Please convert 'string' to 'number' or change variable's type.\\",
      \\"fixed\\": false,
      \\"nodeKind\\": \\"Identifier\\",
      \\"nodeText\\": \\"a\\",
      \\"nodeLocation\\": {
        \\"startLine\\": 1,
        \\"startColumn\\": 10,
        \\"endLine\\": 1,
        \\"endColumn\\": 20
      }
    },
    {
      \\"analysisTarget\\": \\"/base/path/file1.ts\\",
      \\"files\\": {
        \\"/base/path/file1.ts\\": {
          \\"fileName\\": \\"/base/path/file1.ts\\",
          \\"location\\": {
            \\"startLine\\": 1,
            \\"startColumn\\": 10,
            \\"endLine\\": 1,
            \\"endColumn\\": 26
          },
          \\"fixed\\": false,
          \\"hint\\": \\"Argument of type '{0}' is not assignable to parameter of type 'string'. Consider verifying both types, using type assertion: '({} as string)', or using type guard: 'if ({} instanceof string) { ... }'.\\",
          \\"hintAdded\\": true,
          \\"roles\\": [
            \\"analysisTarget\\",
            \\"unmodified\\"
          ]
        }
      },
      \\"errorCode\\": 2345,
      \\"category\\": \\"Error\\",
      \\"message\\": \\"Argument of type '{}' is not assignable to parameter of type 'string'.\\",
      \\"hint\\": \\"Argument of type '{0}' is not assignable to parameter of type 'string'. Consider verifying both types, using type assertion: '({} as string)', or using type guard: 'if ({} instanceof string) { ... }'.\\",
      \\"fixed\\": false,
      \\"nodeKind\\": \\"ObjectLiteralExpression\\",
      \\"nodeText\\": \\"{}\\",
      \\"nodeLocation\\": {
        \\"startLine\\": 1,
        \\"startColumn\\": 10,
        \\"endLine\\": 1,
        \\"endColumn\\": 26
      }
    }
  ]
}"
`;

exports[`Test reporter > print, pull-request-md 1`] = `
"### Summary:
Typescript Version: 4.7.4
Files Tested: 1

### Results:

#### File: /file1.ts, issues: 3:

**Error TS6133**: FIXED
The declaration 'react' is never read or used. Remove the declaration or use it.
Code: \`import react from 'react';\`

**Error TS2322**: NEED TO BE FIXED MANUALLY
The variable 'a' has type 'number', but 'string' is assigned. Please convert 'string' to 'number' or change variable's type.
Code: \`a\`

**Error TS2345**: NEED TO BE FIXED MANUALLY
Argument of type '{0}' is not assignable to parameter of type 'string'. Consider verifying both types, using type assertion: '({} as string)', or using type guard: 'if ({} instanceof string) { ... }'.
Code: \`{}\`
"
`;

exports[`Test reporter > save 1`] = `
"{
  \\"summary\\": {
    \\"projectName\\": \\"@rehearsal/test\\",
    \\"tsVersion\\": \\"4.7.4\\",
    \\"timestamp\\": \\"9/16/2022, 13:24:57\\",
    \\"basePath\\": \\"/base/path\\",
    \\"commandName\\": \\"@rehearsal/reporter\\"
  },
  \\"items\\": [
    {
      \\"analysisTarget\\": \\"/base/path/file1.ts\\",
      \\"files\\": {
        \\"/base/path/file1.ts\\": {
          \\"fileName\\": \\"/base/path/file1.ts\\",
          \\"location\\": {
            \\"startLine\\": 1,
            \\"startColumn\\": 1,
            \\"endLine\\": 1,
            \\"endColumn\\": 1
          },
          \\"fixed\\": true,
          \\"code\\": \\"import react from 'react';\\",
          \\"codeFixAction\\": \\"delete\\",
          \\"hintAdded\\": false,
          \\"roles\\": [
            \\"analysisTarget\\",
            \\"modified\\"
          ]
        }
      },
      \\"errorCode\\": 6133,
      \\"category\\": \\"Error\\",
      \\"message\\": \\"'react' is declared but its value is never read.\\",
      \\"hint\\": \\"The declaration 'react' is never read or used. Remove the declaration or use it.\\",
      \\"fixed\\": true,
      \\"nodeKind\\": \\"ImportDeclaration\\",
      \\"nodeText\\": \\"import react from 'react';\\",
      \\"nodeLocation\\": {
        \\"startLine\\": 1,
        \\"startColumn\\": 1,
        \\"endLine\\": 1,
        \\"endColumn\\": 12
      }
    },
    {
      \\"analysisTarget\\": \\"/base/path/file1.ts\\",
      \\"files\\": {
        \\"/base/path/file1.ts\\": {
          \\"fileName\\": \\"/base/path/file1.ts\\",
          \\"location\\": {
            \\"startLine\\": 1,
            \\"startColumn\\": 10,
            \\"endLine\\": 1,
            \\"endColumn\\": 20
          },
          \\"fixed\\": false,
          \\"hint\\": \\"The variable 'a' has type 'number', but 'string' is assigned. Please convert 'string' to 'number' or change variable's type.\\",
          \\"hintAdded\\": true,
          \\"roles\\": [
            \\"analysisTarget\\",
            \\"unmodified\\"
          ]
        }
      },
      \\"errorCode\\": 2322,
      \\"category\\": \\"Error\\",
      \\"message\\": \\"Type 'string' is not assignable to type 'number'.\\",
      \\"hint\\": \\"The variable 'a' has type 'number', but 'string' is assigned. Please convert 'string' to 'number' or change variable's type.\\",
      \\"fixed\\": false,
      \\"nodeKind\\": \\"Identifier\\",
      \\"nodeText\\": \\"a\\",
      \\"nodeLocation\\": {
        \\"startLine\\": 1,
        \\"startColumn\\": 10,
        \\"endLine\\": 1,
        \\"endColumn\\": 20
      }
    },
    {
      \\"analysisTarget\\": \\"/base/path/file1.ts\\",
      \\"files\\": {
        \\"/base/path/file1.ts\\": {
          \\"fileName\\": \\"/base/path/file1.ts\\",
          \\"location\\": {
            \\"startLine\\": 1,
            \\"startColumn\\": 10,
            \\"endLine\\": 1,
            \\"endColumn\\": 26
          },
          \\"fixed\\": false,
          \\"hint\\": \\"Argument of type '{0}' is not assignable to parameter of type 'string'. Consider verifying both types, using type assertion: '({} as string)', or using type guard: 'if ({} instanceof string) { ... }'.\\",
          \\"hintAdded\\": true,
          \\"roles\\": [
            \\"analysisTarget\\",
            \\"unmodified\\"
          ]
        }
      },
      \\"errorCode\\": 2345,
      \\"category\\": \\"Error\\",
      \\"message\\": \\"Argument of type '{}' is not assignable to parameter of type 'string'.\\",
      \\"hint\\": \\"Argument of type '{0}' is not assignable to parameter of type 'string'. Consider verifying both types, using type assertion: '({} as string)', or using type guard: 'if ({} instanceof string) { ... }'.\\",
      \\"fixed\\": false,
      \\"nodeKind\\": \\"ObjectLiteralExpression\\",
      \\"nodeText\\": \\"{}\\",
      \\"nodeLocation\\": {
        \\"startLine\\": 1,
        \\"startColumn\\": 10,
        \\"endLine\\": 1,
        \\"endColumn\\": 26
      }
    }
  ]
}"
`;
